tests/integration/juju.py

```diff
@@ -0,0 +1,100 @@
+#!/usr/bin/env python3
+# Copyright 2022 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+
+import json
+import subprocess
+import time
+from typing import Dict, List
+
+
+class Juju:
+    @classmethod
+    def model_name(cls):
+        return cls.status()["model"]["name"]
+
+    @classmethod
+    def status(cls):
+        args = ["status", "--format", "json"]
+        result = cls.cli(*args)
+        return json.loads(result.stdout)
+
+    @classmethod
+    def deploy(
+        cls,
+        charm: str,
+        *,
+        alias: str | None = None,
+        channel: str | None = None,
+        config: Dict[str, str] = {},
+        resources: Dict[str, str] = {},
+        trust: bool = False,
+    ):
+        args = ["deploy", charm]
+
+        if alias:
+            args = [*args, alias]
+
+        if channel:
+            args = [*args, "--channel", channel]
+
+        if config:
+            for k, v in config.items():
+                args = [*args, "--config", f"{k}={v}"]
+
+        if resources:
+            for k, v in resources.items():
+                args = [*args, "--resource", f"{k}={v}"]
+
+        if trust:
+            args = [*args, "--trust"]
+
+        return cls.cli(*args)
+
+    @classmethod
+    def integrate(cls, requirer: str, provider: str):
+        args = ["integrate", requirer, provider]
+        return cls.cli(*args)
+
+    @classmethod
+    def run(cls, unit: str, action: str):
+        args = ["run", "--format", "json", unit, action]
+        act = cls.cli(*args)
+        result = json.loads(act.stdout)
+        return result[unit]["results"]
+
+    @classmethod
+    def wait_for_idle(cls, applications: List[str], timeout: int):
```

Comment:
python-libjuju's `wait_for_idle` is one of the few places where that package adds value for most users, mainly in the `raise_on_blocked` and `raise_on_error` options.  If this gets used more widely, those are important features to include for catching transient behaviour that shouldn't occur but can get overlooked